rllib/core/models/specs/specs_base.py

```diff
@@ -87,26 +95,40 @@ class TensorSpec(Spec):
         validate: Checks if the shape and dtype of the tensor matches the
             specification.
         fill: creates a tensor with the specified value that is an
-            example of a tensor that matches the specification.
-
-    Abstract Methods:
-        get_type: Returns the type of the tensor, e.g. tf.Tensor or torch.Tensor.
-        get_shape: Returns the shape of the tensor depending on the backend.
-        get_dtype: Returns the dtype of the tensor depending on the backend.
-        _full: Creates a tensor with the specified value that
-            has values of fill_value, shape of shape, and dtype of self.dtype.
+            example of a tensor that matches the specification. This can only be
+            called if `framework` is specified.
     """
 
     def __init__(
         self,
         shape: str,
         *,
         dtype: Optional[Any] = None,
+        framework: Optional[str] = None,
         **shape_vals: int,
     ) -> None:
         self._expected_shape = self._parse_expected_shape(shape, shape_vals)
         self._full_shape = self._get_full_shape()
         self._dtype = dtype
+        self._framework = framework
+
+        if framework not in ("tf2", "torch", "np", "jax", None):
+            raise ValueError(f"Unknown framework {self._framework}")
+
+    @property
+    def _expected_type(self) -> Type:
+        if self._framework == "torch":
+            return torch.Tensor
+        elif self._framework == "tf2":
+            return tf.Tensor
+        elif self._framework == "np":
+            return np.ndarray
+        elif self._framework == "jax":
+            jax, _ = try_import_jax()
+            return jax.numpy.ndarray
+        elif self._framework is None:
+            # Don't restrict the type of the tensor if no framework is specified.
+            return object
```

Comment:
Before this PR, we'd extend this base class. Which made adding new frameworks very easy.
Now the pattern to add a framework to this (which I hope we will more or less never have to do) is to change this class.